File: DataStructureAndAlgorithm/DataStructures/Graphs/Python/graph.py
from collections import defaultdict, deque

class Graph:
    def __init__(self, directed=False):
        self.graph = defaultdict(list)
        self.directed = directed

    def add_edge(self, u, v):
        self.graph[u].append(v)
        if not self.directed:
            self.graph[v].append(u)

    def print_graph(self):
        for vertex in self.graph:
            print(f"Vertex {vertex}:", " -> ".join(map(str, self.graph[vertex])))

    def bfs(self, start_vertex):
        visited = {start_vertex}
        queue = deque([start_vertex])
        while queue:
            vertex = queue.popleft()
            print(f"Visited {vertex}")
            for neighbour in self.graph[vertex]:
                if neighbour not in visited:
                    visited.add(neighbour)
                    queue.append(neighbour)

    def dfs(self, start_vertex):
        visited = set()
        self._dfs_util(start_vertex, visited)

    def _dfs_util(self, vertex, visited):
        visited.add(vertex)
        print(f"Visited {vertex}")
        for neighbour in self.graph[vertex]:
            if neighbour not in visited:
                self._dfs_util(neighbour, visited)

File: DataStructureAndAlgorithm/DataStructures/Graphs/C++/Graph.cpp
#include "Graph.h"
#include <vector>

Graph::Graph(int vertices, bool directed) : numVertices(vertices), isDirected(directed) {
    adjLists.resize(vertices);
}

void Graph::addEdge(int src, int dest) {
    adjLists[src].push_back(dest);
    if (!isDirected) {
        adjLists[dest].push_back(src);
    }
}

void Graph::printGraph() {
    for (int i = 0; i < numVertices; ++i) {
        std::cout << "Adjacency list of vertex " << i << ": ";
        for (int vertex : adjLists[i]) {
            std::cout << vertex << " -> ";
        }
        std::cout << "nullptr" << std::endl;
    }
}

void Graph::BFS(int startVertex) {
    std::vector<bool> visited(numVertices, false);
    std::list<int> queue;

    visited[startVertex] = true;
    queue.push_back(startVertex);

    while (!queue.empty()) {
        int currVertex = queue.front();
        std::cout << "Visited " << currVertex << std::endl;
        queue.pop_front();

        for (int adjVertex : adjLists[currVertex]) {
            if (!visited[adjVertex]) {
                visited[adjVertex] = true;
                queue.push_back(adjVertex);
            }
        }
    }
}

void Graph::DFSUtil(int v, std::vector<bool>& visited) {
    visited[v] = true;
    std::cout << "Visited " << v << std::endl;

    for (int adjVertex : adjLists[v]) {
        if (!visited[adjVertex]) {
            DFSUtil(adjVertex, visited);
        }
    }
}

void Graph::DFS(int startVertex) {
    std::vector<bool> visited(numVertices, false);
    DFSUtil(startVertex, visited);
}
